core/unit_test/TestExecSpaceThreadSafety.hpp

```diff
@@ -0,0 +1,537 @@
+//
+//@HEADER
+
+#include <cstdio>
+#include <sstream>
+#include <iostream>
+#include <thread>
+
+#include <Kokkos_Core.hpp>
+
+namespace Test {
+
+#include <Kokkos_Core.hpp>
+#include <iostream>
+#include <thread>
+
+#ifdef KOKKOS_ENABLE_OPENMP
+template <class Lambda1, class Lambda2>
+void run_threaded_test(const Lambda1 l1, const Lambda2 l2) {
+#pragma omp parallel num_threads(2)
+  {
+    if (omp_get_thread_num() == 0) l1();
+    if (omp_get_thread_num() == 1) l2();
+  }
+}
+// We cannot run the multithreaded test when threads or HPX is enabled because
+// we cannot launch a thread from inside another thread
+#elif !defined(KOKKOS_ENABLE_THREADS) && !defined(KOKKOS_ENABLE_HPX)
+template <class Lambda1, class Lambda2>
+void run_threaded_test(const Lambda1 l1, const Lambda2 l2) {
+  std::thread t1(std::move(l1));
+  std::thread t2(std::move(l2));
+  t1.join();
+  t2.join();
+}
+#else
+template <class Lambda1, class Lambda2>
+void run_threaded_test(const Lambda1 l1, const Lambda2 l2) {
+  l1();
+  l2();
+}
+#endif
+
+void run_exec_space_thread_safety_range() {
+  constexpr int N = 1000;
+  constexpr int M = 1000;
+
+  Kokkos::View<int *, TEST_EXECSPACE> view("view", N);
+  Kokkos::View<Kokkos::pair<int, int>, TEST_EXECSPACE> error_index(
+      "error_index");
+
+  run_threaded_test(
+      [=]() {
+        TEST_EXECSPACE exec;
+        for (int j = 0; j < M; ++j) {
+          Kokkos::parallel_for(
+              "bar", Kokkos::RangePolicy<TEST_EXECSPACE>(exec, 0, N),
+              KOKKOS_LAMBDA(int i) {
+                if (i < N - 1 &&
+                    (Kokkos::atomic_load(view.data() + (i + 1)) - view(i)) > 1)
+                  Kokkos::atomic_store(error_index.data(),
+                                       Kokkos::pair{j + 1, i});
+                ++view(i);
+              });
+        }
+        exec.fence();
+      },
+      [=]() {
+        TEST_EXECSPACE exec;
+        for (int j = 0; j < M; ++j) {
+          Kokkos::parallel_for(
+              "foo", Kokkos::RangePolicy<TEST_EXECSPACE>(exec, 0, N),
+              KOKKOS_LAMBDA(int i) {
+                if (i < N - 1 &&
+                    (Kokkos::atomic_load(view.data() + (N - 2 - i)) -
+                     view(N - 1 - i)) > 1)
+                  Kokkos::atomic_store(error_index.data(),
+                                       Kokkos::pair{j + 1, N - 1 - i});
+                ++view(N - 1 - i);
+              });
+        }
+        exec.fence();
+      });
+
+  auto host_error_index =
+      Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace{}, error_index);
+  auto [iteration, element] = Kokkos::pair<int, int>(host_error_index());
+  ASSERT_EQ(iteration, 0) << " failing at " << iteration << ", " << element;
+  auto host_view =
+      Kokkos::create_mirror_view_and_copy(Kokkos::HostSpace{}, view);
+  for (int i = 0; i < N; ++i)
+    ASSERT_EQ(host_view(i), 2 * M) << " failing at " << i;
+}
+
+TEST(TEST_CATEGORY, exec_space_thread_safety_range) {
+// FIXME_OPENACC
```

Comment:
It looks like `OpenACC` does not pass any of the tests, so maybe just remove the test at the CMake level like you did for `Serial` and `OpenMP`

Reply:
Ditto

Reply:
(Or clearly state why these are treated differently)